qgym/benchmarks/metrics/routing_metrics.py

```diff
@@ -0,0 +1,206 @@
+                interaction_fidelities.append(
+                    connection_graph[qubit1][qubit2]["weight"]
+                )
+
+        initial_fidelity_mean = mean(interaction_fidelities)
+
+        for interaction in swaps_gates_inserted:
+            (_, qubit1, qubit2) = interaction
+            if interaction in interactions:
+                index = interactions.index(interaction)
+                interaction_fidelities[index] *= connection_graph[qubit1][qubit2][
+                    "weight"
+                ]
+            else:
+                interactions.append(interaction)
+                interaction_fidelities.append(
+                    connection_graph[qubit1][qubit2]["weight"]
+                )
+
+        final_fidelity_mean = mean(interaction_fidelities)
+
+        return initial_fidelity_mean / final_fidelity_mean
+
+    def average_qubit_fidelity_ratio_loss(
+        self,
+        initial_interaction_circuit: NDArray[np.int_],
+        connection_graph: nx.Graph,
+        final_interaction_circuit: NDArray[np.int_],
+    ) -> int:
+        """Method which calculates the average fidelity on qubits, where the fidelity of
+        one qubit is calculated by consecutively multiplying 1 with the
+        connection_graph_edge_fidelity for those edge in the interaction graph that are
+        connected to that particular qubit. The averages over all qubits before and
+        after adding swaps are calculated and divided to get the ratio.
+
+        Args:
+            initial_interaction_circuit: An array of 2-tuples of integers, where every
+                tuple represents a, not specified, gate acting on the two qubits labeled
+                by the integers in the tuples.
+            connection_graph: :class:`networkx.Graph` representation of the QPU
+                topology. Each node represents a physical qubit and each edge represents
+                a connection in the QPU topology.
+            final_interaction_circuit: An array of 2-tuples of integers, where every
+                tuple represents a, not specified, gate acting on the two qubits labeled
+                by the integers in the tuples.
+        """
+
+        initial_interaction_qubits = []
+        for qubit1, qubit2 in initial_interaction_circuit:
+            if qubit1 not in initial_interaction_qubits:
+                initial_interaction_qubits.append(qubit1)
+            if qubit2 not in initial_interaction_qubits:
+                initial_interaction_qubits.append(qubit2)
+
+        initial_qubit_fidelities = []
+        for qubit in initial_interaction_qubits:
+            qubit_fidelity = 1
+            for qubit1, qubit2 in initial_interaction_circuit:
+                if qubit == qubit1 or qubit == qubit2:
+                    qubit_fidelity *= connection_graph[qubit1][qubit2]["weight"]
+            initial_qubit_fidelities.append(qubit_fidelity)
+        initial_qubit_fidelity_mean = mean(initial_qubit_fidelities)
```

Comment:
All this code is basically a _transform_ of `initial_interaction_qubits` into `initial_qubit_fidelities`, i.e., we walk the former and tranform it into the latter.

And the transformation function  is a _reduction_, i.e., we walk `initial_interaction_circuit` and come up with a fidelity value.

```
initial_qubit_fidelities = [
    functools.reduce(
        lambda fidelity, tuple : fidelity * connection_graph[ t[0] ][ t[1] ][ "weight" ] if q in tuple else fidelity,
        initial_interaction_circuit,
        1
    )
    for q in initial_interaction_qubits
]
```

Transformations can be implemented as list comprehensions in Python.

For reductions, we have `functools.reduce`.

In general, thinking in terms of algorithms instead of for loops:
1. Is more elegant, more concise.
2. Most importantly, it expresses intent better.